contracts/extensions/ERC721AQueries.sol

```diff
@@ -0,0 +1,158 @@
+     *   - `startTimestamp` = `0`
+     *   - `burned` = `false`
+     *
+     * If the `tokenId` is burned:
+     *   - `addr` = `<Address of owner before token was burned>`
+     *   - `startTimestamp` = `<Timestamp when token was burned>`
+     *   - `burned = `true`
+     *
+     * Otherwise:
+     *   - `addr` = `<Address of owner>`
+     *   - `startTimestamp` = `<Timestamp of start of ownership>`
+     *   - `burned = `false`
+     */
+    function rawOwnershipOf(uint256 tokenId) public view returns (TokenOwnership memory) {
+        TokenOwnership memory ownership;
+        if (tokenId < _startTokenId() || tokenId >= _currentIndex) {
+            return ownership;
+        }
+        ownership = _ownerships[tokenId];
+        if (ownership.burned) {
+            return ownership;
+        }
+        return _ownershipOf(tokenId);
+    }
+
+    /**
+     * @dev Returns an array of `TokenOwnership` structs at `tokenIds` in order.
+     * See {ERC721AQueries-rawOwnershipOf}
+     */
+    function rawOwnershipsOf(uint256[] memory tokenIds) external view returns (TokenOwnership[] memory) {
+        unchecked {
+            uint256 tokenIdsLength = tokenIds.length;
+            TokenOwnership[] memory ownerships = new TokenOwnership[](tokenIdsLength);
+            for (uint256 i; i != tokenIdsLength; ++i) {
+                ownerships[i] = rawOwnershipOf(tokenIds[i]);
+            }
+            return ownerships;
+        }
+    }
+
+    /**
+     * @dev Returns an array of token IDs owned by `owner`,
+     * in the range [`start`, `stop`)
+     * (i.e. `start <= tokenId < stop`).
+     *
+     * This function allows for tokens to be queried if the collection
+     * grows too big for a single call of {ERC721AQueries-tokensOfOwner}.
+     */
+    function tokensOfOwnerIn(
+        address owner,
+        uint256 start,
+        uint256 stop
+    ) external view returns (uint256[] memory) {
+        unchecked {
+            uint256 tokenIdsIdx;
+            uint256 stopLimit = _currentIndex;
+            // Set `start = max(start, _startTokenId())`.
+            if (start < _startTokenId()) {
+                start = _startTokenId();
+            }
+            // Set `stop = min(stop, _currentIndex)`.
+            if (stop > stopLimit) {
+                stop = stopLimit;
+            }
+            uint256 tokenIdsMaxLength = balanceOf(owner);
+            // Set `tokenIdsMaxLength = min(balanceOf(owner), stop - start)`,
+            // to cater for cases where `balanceOf(owner)` is too big.
+            if (start < stop) {
+                uint256 windowLength = stop - start;
+                if (windowLength < tokenIdsMaxLength) {
+                    tokenIdsMaxLength = windowLength;
+                }
+            } else {
+                tokenIdsMaxLength = 0;
+            }
+            uint256[] memory tokenIds = new uint256[](tokenIdsMaxLength);
+            if (tokenIdsMaxLength == 0) {
+                return tokenIds;
```

Comment:
Shouldn't you return an empty array here?

Reply:
I want to recycle line 93 to reduce the amount of opcodes. 
This function is external.